In [1]:
import numpy as np
import pandas as pd
import sklearn
import matplotlib.pyplot as plt

### PLAN

1. Pobranie danych
2. Przejrzeć co mamy
    - wielkosc data setu
    - jakie kolumny
    - jakiego typu dane
    - podejrzec probke
    - jesli dane sa we wlasciwym formacie: proste statystyki
3. Konwersja danych na odpowiedni format
4. Co z brakujacymi wartosciami?

#### Zaimportuj i przejrzyj dane

In [2]:
dane = pd.read_csv('data/DataWorkshop.csv')

In [3]:
print('Rozmiar danych: ', dane.shape)

Rozmiar danych:  (229929, 30)


In [4]:
print('Kolumny w datasecie: ')
dane.columns

Kolumny w datasecie: 


Index(['Data zdarzenia', 'Godzina', 'GPS x', 'GPS y', 'Miejscowość', 'Ulica',
       'Numer domu', 'Numer drogi', 'KM HM (Pikietaż)', 'Obszar',
       'Skrzyżowanie', 'Geometria drogi', 'Charakterystyka miejsca zdarzenia',
       'Oświetlenie', 'Warunki atmosferyczne', 'Rodzaj zdarzenia', 'Wina',
       'Rodzaj drogi', 'Nawierzchnia drogi', 'Stan nawierzchni',
       'Sygnalizacja świetlna', 'Oznakowanie poziome',
       'Liczba pasów w zdarzeniu', 'Zachowanie kierującego',
       'Zachowanie pieszego', 'Inne przyczyny', 'Liczba rannych w zdarzeniu',
       'Liczba zabitych w zdarzeniu', 'Liczba kolizji', 'Liczba wypadków'],
      dtype='object')

In [5]:
print('Typy danych: ')
dane.dtypes

Typy danych: 


Data zdarzenia                        object
Godzina                               object
GPS x                                 object
GPS y                                 object
Miejscowość                           object
Ulica                                 object
Numer domu                            object
Numer drogi                           object
KM HM (Pikietaż)                      object
Obszar                                object
Skrzyżowanie                          object
Geometria drogi                       object
Charakterystyka miejsca zdarzenia     object
Oświetlenie                           object
Warunki atmosferyczne                 object
Rodzaj zdarzenia                      object
Wina                                  object
Rodzaj drogi                          object
Nawierzchnia drogi                    object
Stan nawierzchni                      object
Sygnalizacja świetlna                 object
Oznakowanie poziome                   object
Liczba pas

In [6]:
print('Brakujace wartosci: ')

missing = dane.isnull().sum()
print(missing)

Brakujace wartosci: 
Data zdarzenia                            0
Godzina                                   0
GPS x                                     0
GPS y                                     0
Miejscowość                               0
Ulica                                 20728
Numer domu                            93884
Numer drogi                               0
KM HM (Pikietaż)                     194888
Obszar                                    1
Skrzyżowanie                         166779
Geometria drogi                       66008
Charakterystyka miejsca zdarzenia         2
Oświetlenie                               0
Warunki atmosferyczne                     1
Rodzaj zdarzenia                          0
Wina                                      0
Rodzaj drogi                              1
Nawierzchnia drogi                        1
Stan nawierzchni                          1
Sygnalizacja świetlna                     0
Oznakowanie poziome                       2
Liczba pasó

In [7]:
print('Brakujace wartosci jako procent zbioru, %: ')
index=[]
lst=[]
for item in missing.iteritems():
    index.append(item[0])
    lst.append(round((item[1]/dane.shape[0])*100,4))
missing_percentage=pd.Series(lst, index=index)
print(missing_percentage)

Brakujace wartosci jako procent zbioru, %: 
Data zdarzenia                        0.0000
Godzina                               0.0000
GPS x                                 0.0000
GPS y                                 0.0000
Miejscowość                           0.0000
Ulica                                 9.0150
Numer domu                           40.8317
Numer drogi                           0.0000
KM HM (Pikietaż)                     84.7601
Obszar                                0.0004
Skrzyżowanie                         72.5350
Geometria drogi                      28.7080
Charakterystyka miejsca zdarzenia     0.0009
Oświetlenie                           0.0000
Warunki atmosferyczne                 0.0004
Rodzaj zdarzenia                      0.0000
Wina                                  0.0000
Rodzaj drogi                          0.0004
Nawierzchnia drogi                    0.0004
Stan nawierzchni                      0.0004
Sygnalizacja świetlna                 0.0000
Oznakowanie

In [8]:
# Probka danych
dane.sample(50)

Data zdarzenia   Godzina      GPS x      GPS y           Miejscowość  \
147450     2018-08-31  13:15:00  19*08'174  50*16'504             SOSNOWIEC   
111843     2018-01-08  08:50:00  19*04'561  50*49'219           CZĘSTOCHOWA   
12398      2016-04-07  04:55:00  19*00'417  49*53'486  CZECHOWICE-DZIEDZICE   
183411     2019-04-01  17:37:00  18*58'176  50*17'571               CHORZÓW   
18290      2016-05-17  14:15:00  19*06'390  50*17'250             SOSNOWIEC   
28769      2016-07-30  14:00:00  18*38'597  50*18'394               GLIWICE   
207584     2019-08-28  11:00:00  18*47'033  49*47'316               SKOCZÓW   
203456     2019-08-01  17:35:00  19*07'584  50*18'010             SOSNOWIEC   
88359      2017-08-28  08:57:00  19*10'466  50*04'327                BIERUŃ   
67860      2017-04-12  17:30:00  19*10'273  50*17'250             SOSNOWIEC   
18068      2016-05-14  13:55:00  18*57'374  50*23'409       PIEKARY ŚLĄSKIE   
157095     2018-10-25  09:35:00  19*01'149  50*10'136                 TYCHY   
115397     2018-01-31  16:25:00  18*46'300  49*48'297               SKOCZÓW   
89718      2017-08-27  17:40:00  18*43'500  49*50'468                USTROŃ   
110127     2017-12-23  08:00:00  18*43'097  50*15'587               GLIWICE   
64794      2017-03-23  20:30:00  19*31'124  50*27'055          OGRODZIENIEC   
173220     2019-01-28  10:00:00  18*58'480  50*20'443       PIEKARY ŚLĄSKIE   
143957     2018-08-07  15:50:00  19*10'192  50*18'020             SOSNOWIEC   
68189      2017-04-14  14:35:00  18*46'451  50*18'000                ZABRZE   
203696     2019-08-01  17:45:00  19*15'361  50*22'381      DĄBROWA GÓRNICZA   
106042     2017-12-04  16:00:00  19*07'489  50*16'266             SOSNOWIEC   
76803      2017-06-09  20:10:00  18*41'377  50*15'137               GLIWICE   
192679     2019-05-27  10:40:00  19*11'477  50*54'398               MYKANÓW   
118280     2018-02-19  17:30:00  18*58'226  50*13'545              KATOWICE   
105811     2017-12-02  12:50:00  19*07'441  50*16'268             SOSNOWIEC   
20222      2016-05-30  16:10:00  19*03'218  49*48'282         BIELSKO-BIAŁA   
7730       2016-03-01  14:00:00  18*59'061  50*07'267                 TYCHY   
90025      2017-09-06  15:00:00  18*57'066  50*22'224       PIEKARY ŚLĄSKIE   
193183     2019-05-29  05:45:00  19*16'284  50*14'414              JAWORZNO   
15012      2016-04-22  22:00:00  18*57'136  50*17'421               CHORZÓW   
174538     2019-02-04  22:20:00  19*02'531  50*15'506              KATOWICE   
150539     2018-09-18  14:47:00  18*55'350  49*59'064              PSZCZYNA   
180253     2019-03-12  20:10:00  19*04'276  49*43'439            BUCZKOWICE   
161521     2018-11-22  21:00:00  19*00'103  50*18'016  SIEMIANOWICE ŚLĄSKIE   
32310      2016-08-25  14:45:00  18*58'539  50*07'180                 TYCHY   
54955      2017-01-12  12:15:00  18*47'084  50*19'396                ZABRZE   
39837      2016-10-12  14:30:00  19*13'551  50*19'026      DĄBROWA GÓRNICZA   
137712     2018-06-25  12:40:00  19*13'569  50*09'245               IMIELIN   
44959      2016-11-13  14:30:00  18*42'389  50*00'082            WARSZOWICE   
11867      2016-03-31  20:50:00  18*39'334  50*18'054               GLIWICE   
154503     2018-10-11  16:42:00  18*59'131  50*07'259                 TYCHY   
71493      2017-05-09  09:10:00  18*38'383  49*44'275               CIESZYN   
45492      2016-11-17  00:05:00  18*51'129  50*12'255               MIKOŁÓW   
132274     2018-05-23  14:40:00  19*13'473  50*20'228      DĄBROWA GÓRNICZA   
67004      2017-04-07  11:48:00  19*08'571  50*46'515           CZĘSTOCHOWA   
108050     2017-12-15  03:24:00  19*10'307  50*33'388             MARKOWICE   
33484      2016-08-31  13:40:00  19*04'533  49*48'308         BIELSKO-BIAŁA   
190506     2019-05-14  19:50:00  18*39'376  50*18'170               GLIWICE   
51170      2016-12-18  06:00:00  18*42'402  50*15'176               GLIWICE   
140865     2018-07-18  09:30:00  18*53'360  49*51'233    

### Konwersja typu danych

#### Data i czas

In [9]:
# Utworzenie kopii zbioru danych. Dalsza konwersja odbywac sie bedzie na kopii danych
df=dane.copy()

In [10]:
# Data zdarzenia i Godzina scalone do jednej kolumny
print('kolumna DATA ZDARZENIA: ', df['Data zdarzenia'].dtypes)
print('kolumna GODZINA: ', df['Data zdarzenia'].dtypes)
df['data_full'] = df['Data zdarzenia'] + ' ' + df['Godzina']
df.sample(10)

kolumna DATA ZDARZENIA:  object
kolumna GODZINA:  object


Data zdarzenia   Godzina      GPS x      GPS y    Miejscowość  \
111923     2018-01-08  17:14:00  19*00'501  50*15'370       KATOWICE   
168260     2018-12-29  13:25:00  19*03'063  50*14'385       KATOWICE   
45225      2016-11-15  17:00:00  18*47'266  49*44'098         USTROŃ   
161909     2018-11-24  15:05:00  18*46'584  50*17'347         ZABRZE   
78256      2017-06-20  09:07:00  18*58'539  50*14'102       KATOWICE   
69543      2017-04-25  19:25:00  18*57'425  50*45'291    ALEKSANDRIA   
214868     2019-10-09  13:20:00  18*39'445  50*18'163        GLIWICE   
75433      2017-06-01  18:00:00  19*13'483  50*27'591       SIEWIERZ   
131378     2018-05-16  19:07:00  19*13'435  50*36'188    LGOTA GÓRNA   
135886     2018-06-15  13:59:00  19*02'156  49*50'227  BIELSKO-BIAŁA   

                           Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
111923                  SOKOLSKA        NaN      P6519S              NaN   
168260                       NaN   NIEZNANE          A4            340,3   
45225                  KATOWICKA        NaN        W941              5,9   
161909   GEN.CHARLESA DE GAULLEA        107      P8204S              NaN   
78256                    LIGOCKA         68      P6465S              NaN   
69543                        NaN        NaN        W904              NaN   
214868  DROGOWA TRASA ŚREDNICOWA        NaN        W902              NaN   
75433                        NaN        NaN           1              NaN   
131378                   ŻARECKA        NaN        W789              NaN   
135886                WARSZAWSKA        154      P7477S              NaN   

                      Obszar  ... Oznakowanie poziome  \
111923     Obszar zabudowany  ...                Jest   
168260  Obszar niezabudowany  ...                Jest   
45225   Obszar niezabudowany  ...                Jest   
161909     Obszar zabudowany  ...                Jest   
78256      Obszar zabudowany  ...                Jest   
69543   Obszar niezabudowany  ...                Jest   
214868  Obszar niezabudowany  ...                Jest   
75433   Obszar niezabudowany  ...                Jest   
131378     Obszar zabudowany  ...                Jest   
135886     Obszar zabudowany  ...                Jest   

       Liczba pasów w zdarzeniu  \
111923                      4.0   
168260                      3.0   
45225                       4.0   
161909                      3.0   
78256                       NaN   
69543                       NaN   
214868                      6.0   
75433                       NaN   
131378                      2.0   
135886                      3.0   

                                   Zachowanie kierującego Zachowanie pieszego  \
111923  Nieustąpienie pierwszeństwa pieszemu na przejś...                 NaN   
168260        Niedostosowanie prędkości do warunków ruchu                 NaN   
45225               Nieustąpienie pierwszeństwa przejazdu                 NaN   
161909        Niezachowanie bezp. odleg. między pojazdami                 NaN   
78256               Nieustąpienie pierwszeństwa przejazdu                 NaN   
69543                                                 NaN                 NaN   
214868        Niezachowanie bezp. odleg. między pojazdami                 NaN   
75433               Nieustąpienie pierwszeństwa przejazdu                 NaN   
131378                                     Inne przyczyny                 NaN   
135886                                     Inne przyczyny                 NaN   

                      Inne przyczyny Liczba rannych w zdarzeniu  \
111923                           NaN                          0   
168260                           NaN                          1   
45225                            NaN                          0   
161909                           NaN                          0   
78256                            NaN                          0   
69543   Obiekty, zwierzęta na drodze                          0   
21

In [11]:
# Data zdarzenia i Godzina do typu datetime
df['data_full'] = pd.to_datetime(df['data_full'], format='%Y-%m-%d %H:%M:%S')
df.sample(10)
# TO-DO: Jaki to dzien tygodnia? Jaki to dzien miesiaca? To dzien czy noc? Jaka faza ksiezyca itd..

Data zdarzenia   Godzina      GPS x      GPS y       Miejscowość  \
201311     2019-07-18  14:48:00  19*33'395  50*31'088         ZAWIERCIE   
122699     2018-03-20  18:25:00  19*06'379  50*47'392       CZĘSTOCHOWA   
229766     2019-12-31  17:40:00  18*31'237  50*22'238          TACISZÓW   
161771     2018-11-23  16:32:00  19*04'400  50*48'575       CZĘSTOCHOWA   
116077     2018-01-28  19:05:00  18*42'409  50*15'595           GLIWICE   
178738     2019-03-04  14:00:00  19*22'585  50*39'460        CZATACHOWA   
136253     2018-06-18  15:15:00  19*07'049  50*48'258       CZĘSTOCHOWA   
118240     2018-02-19  11:15:00  18*36'004  49*55'373  JASTRZĘBIE-ZDRÓJ   
143290     2018-08-03  08:05:00  19*06'218  50*49'016       CZĘSTOCHOWA   
19551      2016-05-27  15:30:00  18*42'314  50*18'129           GLIWICE   

                                Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
201311                       JURAJSKA        NaN          78            131,6   
122699  ALEJA BOHATERÓW MONTE CASSINO         40          43              NaN   
229766                       GLIWICKA         12      P2913S              NaN   
161771                   WYSZYŃSKIEGO        109    G111018S              NaN   
116077                        OKRĘŻNA         32          I2              NaN   
178738                            NaN        NaN        W793              NaN   
136253                 ALEJA WOLNOŚCI        NaN      P6614S              NaN   
118240                     CIESZYŃSKA        107        W937              4,6   
143290            ALEJA JANA PAWŁA II        106          46              NaN   
19551                      CHORZOWSKA         52      P7207S              NaN   

                      Obszar  ... Oznakowanie poziome  \
201311  Obszar niezabudowany  ...                Jest   
122699     Obszar zabudowany  ...                Jest   
229766     Obszar zabudowany  ...              Nie ma   
161771     Obszar zabudowany  ...              Nie ma   
116077     Obszar zabudowany  ...              Nie ma   
178738     Obszar zabudowany  ...              Nie ma   
136253     Obszar zabudowany  ...                Jest   
118240     Obszar zabudowany  ...                Jest   
143290     Obszar zabudowany  ...                Jest   
19551      Obszar zabudowany  ...                Jest   

       Liczba pasów w zdarzeniu  \
201311                      2.0   
122699                      4.0   
229766                      2.0   
161771                      2.0   
116077                      1.0   
178738                      2.0   
136253                      4.0   
118240                      2.0   
143290                      7.0   
19551                       2.0   

                                   Zachowanie kierującego Zachowanie pieszego  \
201311                            Nieprawidłowe skręcanie                 NaN   
122699        Niezachowanie bezp. odleg. między pojazdami                 NaN   
229766                                                NaN                 NaN   
161771                                                NaN                 NaN   
116077                              Nieprawidłowe cofanie                 NaN   
178738                                                NaN                 NaN   
136253  Nieustąpienie pierwszeństwa pieszemu na przejś...                 NaN   
118240        Niezachowanie bezp. odleg. między pojazdami                 NaN   
143290        Niezachowanie bezp. odleg. między pojazdami                 NaN   
19551                             Nieprawidłowe: omijanie                 NaN   

                      Inne przyczyny Liczba rannych w zdarzeniu  \
201311                           NaN                          2   
122699                           NaN                          0   
229766  Obiekty, zwierzęta na drodze                          0   
161771                   Stan jezdni                          0   
116077                           NaN            

#### Wspolrzedne geograficzne

Domyslny zapis wspolrzednych to D\*M'S([D]egree, [M]inute, [S]cond) np. 18\*53'289. Jednak, gdy pierwsza cyfra sekund jest zero, czasem wystepuje zapis D\*M''S np. 18\*53''11. Wystepowaly takze bledne zapisy takie jak 4\*\*56'569 i 50\*09'\*01.

In [12]:
def dmsToDecimal(coordinate_as_string):
    try:
        # Zapis domyslny to D*M'S([D]egree, [M]inute, [S]cond) zeby sprawnie podzielic kazda wspolrzedna na liste trzyelementowa
        # zamieniam ' na *, a nastepnie dziele orzymany rezultat znakami *. W trakcie konwersji danych napotkalam rozne nieprawidlowe
        # zapisy opisane w komorce powyzej. Wszystkie wystepujace nietypowe znaki takie jak **, '* i '' zamieniam na * i dziele znakiem *
        degree, minute, second = str(coordinate_as_string).replace("**", "*").replace("'*", "*").replace("''", "*").replace("'", '*').split("*")
    except:
        # Jesli nie dzieli na liste trzyelementowa to najprawdopodobniej trzeci element jest pusty(lista dwuelementowa).
        # Wtedy pierwszy element to stopien, drugi minuta. Sekundy zas sa rowne 0.
        degree, minute = str(coordinate_as_string).replace("**", "*").replace("'*", "*").replace("''", "*").replace("'", '*').split("*")
        second=0
    # W niektorych wypadkach udaje sie podzielic wsadowego stringa tak aby uzyskac ostatnia czesc jako pusty string.
    #Ponizej przeksztalcam pustego stringa w wartosc 0
    if second == "" or second == " ":
        second=0
    # Wlasciwe przeksztalcenie wspolrzednej z wartosci katowych na wartosc liczbowa
    coordinate_as_float = float(degree) + float(minute)/60 + float(second)/(60*60)
    return coordinate_as_float
# Przygotowana wyzej funkcje wywoluje na calych kolumnach dancych
df['GPS x'] = df['GPS x'].apply(lambda x: dmsToDecimal(x))
df['GPS y'] = df['GPS y'].apply(lambda x: dmsToDecimal(x))
df.sample(10)
# TO-DO Wyszukac i wykorzystac biblioteke okreslajaca wspolrzedne GPS na podstawie adresu

Data zdarzenia   Godzina      GPS x      GPS y    Miejscowość  \
78206      2017-06-19  16:10:00  18.767778  49.980833  KOŃCZYCE MAŁE   
47728      2016-11-30  10:38:00  19.274167  50.598889      MARKOWICE   
221607     2019-11-14  15:20:00  19.238889  50.871389    CZĘSTOCHOWA   
177962     2019-02-26  15:15:00  19.115833  50.832778    CZĘSTOCHOWA   
89636      2017-09-04  12:34:00  19.133056  50.916944    CZĘSTOCHOWA   
162478     2018-11-29  00:10:00  19.166111  50.353889       KATOWICE   
125909     2018-04-12  11:13:00  18.953333  50.367778        CHORZÓW   
98907      2017-10-25  07:10:00  18.594444  50.080278         RADLIN   
174527     2019-02-04  17:50:00  19.270000  50.873611    CZĘSTOCHOWA   
159904     2018-11-11  20:20:00  19.425278  49.736389       OKRAJNIK   

                         Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
78206             JAGIELLOŃSKA        NaN        W937               15   
47728                      NaN        NaN           1            506,2   
221607                  ŚLĄSKA        NaN      P6608S              NaN   
177962              ŚW.JADWIGI        5/7          43              NaN   
89636           ALEJA WOLNOŚCI        1/3      P6614S              NaN   
162478             GÓRNOŚLĄSKA         40    G100446S              NaN   
125909               KATOWICKA        114          79              NaN   
98907                 RYBNICKA   NIEZNANE          78             21,5   
174527  ALEJA WOJSKA POLSKIEGO        207          I2              NaN   
159904                ŻYWIECKA          1         946             13,1   

                      Obszar  ... Oznakowanie poziome  \
78206      Obszar zabudowany  ...                Jest   
47728   Obszar niezabudowany  ...                Jest   
221607     Obszar zabudowany  ...                Jest   
177962     Obszar zabudowany  ...                Jest   
89636      Obszar zabudowany  ...                Jest   
162478     Obszar zabudowany  ...                Jest   
125909     Obszar zabudowany  ...                Jest   
98907      Obszar zabudowany  ...                Jest   
174527     Obszar zabudowany  ...                Jest   
159904     Obszar zabudowany  ...                Jest   

       Liczba pasów w zdarzeniu  \
78206                       2.0   
47728                       NaN   
221607                      2.0   
177962                      4.0   
89636                       NaN   
162478                      2.0   
125909                      4.0   
98907                       3.0   
174527                      2.0   
159904                      2.0   

                                   Zachowanie kierującego Zachowanie pieszego  \
78206         Niezachowanie bezp. odleg. między pojazdami                 NaN   
47728         Niezachowanie bezp. odleg. między pojazdami                 NaN   
221607  Nieustąpienie pierwszeństwa pieszemu na przejś...                 NaN   
177962                Nieprawidłowe zmienianie pasa ruchu                 NaN   
89636                                                 NaN                 NaN   
162478              Nieustąpienie pierwszeństwa przejazdu                 NaN   
125909        Niezachowanie bezp. odleg. między pojazdami                 NaN   
98907               Nieustąpienie pierwszeństwa przejazdu                 NaN   
174527                             Nieprawidłowe omijanie                 NaN   
159904                                                NaN                 NaN   

                      Inne przyczyny Liczba rannych w zdarzeniu  \
78206                            NaN                          1   
47728                            NaN                          0   
221607                           NaN                          1   
177962                           NaN                          0   
89636                           Inne                          0   
162478                           NaN                          0   
125909                  

#### Miejsce zdarzenia



In [13]:
# Zmiana nazwy kolumn na nazwy bez spacji i polskich znakow
df.rename(columns={'Miejscowość': 'Miejscowosc', 'Numer domu': "NrDomu", "Numer drogi": "NrDrogi"}, inplace=True)
df.columns

Index(['Data zdarzenia', 'Godzina', 'GPS x', 'GPS y', 'Miejscowosc', 'Ulica',
       'NrDomu', 'NrDrogi', 'KM HM (Pikietaż)', 'Obszar', 'Skrzyżowanie',
       'Geometria drogi', 'Charakterystyka miejsca zdarzenia', 'Oświetlenie',
       'Warunki atmosferyczne', 'Rodzaj zdarzenia', 'Wina', 'Rodzaj drogi',
       'Nawierzchnia drogi', 'Stan nawierzchni', 'Sygnalizacja świetlna',
       'Oznakowanie poziome', 'Liczba pasów w zdarzeniu',
       'Zachowanie kierującego', 'Zachowanie pieszego', 'Inne przyczyny',
       'Liczba rannych w zdarzeniu', 'Liczba zabitych w zdarzeniu',
       'Liczba kolizji', 'Liczba wypadków', 'data_full'],
      dtype='object')

In [14]:
# Pikietaz okresla miejsce zdarzenia na drodze poza terenem zabudowanym
#Zapisywany jako xx,yyy gdzie:
#xx – całkowita liczba kilometrów od początku drogi;
#yyy – całkowita liczba metrów liczona od ostatniego pełnego kilometra.
# W zbiorze danych prawdopodobnie jako wartosc w km, tyle ze z uzyciem "," jako separatora dziesietnego
def pikietazToFloat(pikietazRaw):
    if isinstance(pikietazRaw, str):
        pikietazAsFloat = float(pikietazRaw.replace(",", "."))
    else:
        pikietazAsFloat = pikietazRaw
    return pikietazAsFloat
df['KM HM (Pikietaż)']=df['KM HM (Pikietaż)'].apply(lambda x: pikietazToFloat(x))
df.rename(columns={'KM HM (Pikietaż)': 'KM_HM_Pikietaz'})
df.sample(10)

Data zdarzenia   Godzina      GPS x      GPS y         Miejscowosc  \
189792     2019-05-10  15:42:00  18.695000  50.690000           LUBLINIEC   
136285     2018-06-18  18:02:00  19.518333  50.592778           ZAWIERCIE   
212787     2019-09-27  15:25:00  18.926944  50.442778              ZABRZE   
123529     2018-03-27  01:30:00  19.082778  50.543889  MIASTECZKO ŚLĄSKIE   
206885     2019-08-24  19:00:00  18.798056  50.311667             GLIWICE   
116768     2018-02-09  20:16:00  19.036667  49.805833       BIELSKO-BIAŁA   
169007     2019-01-03  16:30:00  19.148333  50.312500            KATOWICE   
127774     2018-04-23  18:50:00  19.014444  50.296667      ŚWIĘTOCHŁOWICE   
181311     2019-03-19  17:04:00  19.217500  50.808056         CZĘSTOCHOWA   
32         2016-01-03  09:00:00  19.164167  49.679444              LIPOWA   

                                Ulica NrDomu NrDrogi  KM HM (Pikietaż)  \
189792                    MICKIEWICZA    NaN  P2350S               NaN   
136285                   INWESTYCYJNA    NaN  GMINNA               NaN   
212787                       WOLNOŚCI    370  P8206S               NaN   
123529                   STAROWIEJSKA     12  P3256S               NaN   
206885                    TARNOGÓRSKA      1      78               NaN   
116768         GEN.WŁADYSŁAWA ANDERSA     81    W942               NaN   
169007       WALENTEGO ROŹDZIEŃSKIEGO     95      I2               NaN   
127774                        HAJDUKI    NaN  P9803S               NaN   
181311  ALEJA BOHATERÓW MONTE CASSINO     10  P6610S               NaN   
32                                NaN    661  P1405S               5.8   

                   Obszar  ... Oznakowanie poziome Liczba pasów w zdarzeniu  \
189792  Obszar zabudowany  ...                Jest                      1.0   
136285  Obszar zabudowany  ...                Jest                      2.0   
212787  Obszar zabudowany  ...                Jest                      2.0   
123529  Obszar zabudowany  ...              Nie ma                      2.0   
206885  Obszar zabudowany  ...                Jest                      2.0   
116768  Obszar zabudowany  ...                Jest                      3.0   
169007  Obszar zabudowany  ...                Jest                      2.0   
127774  Obszar zabudowany  ...                Jest                      2.0   
181311  Obszar zabudowany  ...                Jest                      4.0   
32      Obszar zabudowany  ...              Nie ma                      2.0   

                             Zachowanie kierującego Zachowanie pieszego  \
189792                       Nieprawidłowe omijanie                 NaN   
136285                                          NaN                 NaN   
212787                               Inne przyczyny                 NaN   
123529  Niedostosowanie prędkości do warunków ruchu                 NaN   
206885        Nieustąpienie pierwszeństwa przejazdu                 NaN   
116768          Nieprawidłowe zmienianie pasa ruchu                 NaN   
169007                               Inne przyczyny                 NaN   
127774                      Nieprawidłowe skręcanie                 NaN   
181311  Niezachowanie bezp. odleg. między pojazdami                 NaN   
32                                              NaN                 NaN   

                      Inne przyczyny Liczba rannych w zdarzeniu  \
189792                           NaN                          0   
136285  Obiekty, zwierzęta na drodze                          0   
212787                           NaN                          0   
123529                           NaN                          0   
206885                           NaN                          0   
116768                           NaN                          0   
169007                           NaN                          0   
127774                           NaN                          0   
181311                           NaN         

In [15]:
# Obszar: zabudowany/niezabudowany
# Przypisane zostaly wartosci logiczne: Obszar zabudowany-True, Obszar niezabudowany False.
# W razie potrzeby takie wartosci mozna wygodnie przeksztalcic na int o wartosciach odpowiednio 1 i 0
df['Obszar'].value_counts()
df['Obszar'] = df['Obszar'].apply(lambda x: True if x=='Obszar zabudowany' else False)
df.sample(10)

Data zdarzenia   Godzina      GPS x      GPS y       Miejscowosc  \
217965     2019-10-25  08:15:00  19.264167  50.921944       CZĘSTOCHOWA   
157689     2018-10-26  21:00:00  18.989444  50.281944             TYCHY   
64730      2017-03-23  09:00:00  18.578333  50.016111  WODZISŁAW ŚLĄSKI   
206389     2019-08-21  09:55:00  19.160278  50.373056          KATOWICE   
124486     2018-04-03  05:40:00  18.665000  50.146667            RYBNIK   
38497      2016-09-29  04:00:00  19.386111  50.958333       RZERZĘCZYCE   
56249      2017-01-22  18:34:00  19.330833  50.296944          JAWORZNO   
197263     2019-06-23  01:20:00  18.274167  50.214722          RACIBÓRZ   
146560     2018-08-25  16:21:00  19.452222  50.599167         ZAWIERCIE   
9189       2016-03-08  07:55:00  18.946111  50.475833             BYTOM   

                         Ulica   NrDomu   NrDrogi  KM HM (Pikietaż)  Obszar  \
217965  ALEJA WOJSKA POLSKIEGO  139/159         1               NaN    True   
157689              MIKOŁOWSKA       28        44              33.0    True   
64730              JASTRZĘBSKA      NaN       933               NaN    True   
206389                SOKOLSKA       15    P6519S               NaN    True   
124486               KOŻDONIÓW      NaN  G180181S               NaN    True   
38497                 MSTOWSKA       30    GMINNA               NaN    True   
56249      KAROLA OLSZEWSKIEGO      NaN  G230281S               NaN    True   
197263                 OPAWSKA      NaN  G290155S               NaN    True   
146560   IGNACEGO DASZYŃSKIEGO      NaN    P1702S               NaN    True   
9189               NARUTOWICZA       22  G150098S               NaN    True   

        ... Oznakowanie poziome Liczba pasów w zdarzeniu  \
217965  ...                Jest                      6.0   
157689  ...                Jest                      3.0   
64730   ...                Jest                      2.0   
206389  ...                Jest                      4.0   
124486  ...              Nie ma                      2.0   
38497   ...                Jest                      NaN   
56249   ...                Jest                      NaN   
197263  ...                Jest                      2.0   
146560  ...                Jest                      2.0   
9189    ...              Nie ma                      NaN   

                             Zachowanie kierującego Zachowanie pieszego  \
217965          Nieprawidłowe zmienianie pasa ruchu                 NaN   
157689          Nieprawidłowe zmienianie pasa ruchu                 NaN   
64730                         Nieprawidłowe cofanie                 NaN   
206389  Niezachowanie bezp. odleg. między pojazdami                 NaN   
124486                                          NaN                 NaN   
38497   Niedostosowanie prędkości do warunków ruchu                 NaN   
56249   Niezachowanie bezp. odleg. między pojazdami                 NaN   
197263                        Nieprawidłowe cofanie                 NaN   
146560        Nieustąpienie pierwszeństwa przejazdu                 NaN   
9189                         Nieprawidłowe: cofanie                 NaN   

                      Inne przyczyny Liczba rannych w zdarzeniu  \
217965                           NaN                          0   
157689                           NaN                          0   
64730                            NaN                          0   
206389                           NaN                          0   
124486  Obiekty, zwierzęta na drodze                          0   
38497                            NaN                          0   
56249                            NaN                          0   
197263                           NaN                          0   
146560                           NaN                          0   
9189                             NaN                          0   

       Liczba zabitych w zdarzeniu Liczba kolizji Liczba wypadków  \
217965           

In [16]:
# Skrzyzowanie
# Zmiana nazwy na niezawierajaca polskich znakow
df['Skrzyżowanie'].unique()
df['Skrzyżowanie'].value_counts()
# Zamiana wartosci tekstowych na int zgodnie z zasada:
# 0 : O ruchu okrężnym; 1 : Równorzędne; 2 : Z drogą z pierwsz.
skrzyzowaniemap = {'O ruchu okrężnym':0, 'Równorzędne':1, 'Z drogą z pierwsz.':2}
df=df.applymap(lambda x: skrzyzowaniemap.get(x) if x in skrzyzowaniemap else x)
df.rename(columns={'Skrzyżowanie': 'Skrzyzowanie'})

Data zdarzenia   Godzina      GPS x      GPS y       Miejscowosc  \
0          2016-01-01  05:15:00  18.658056  50.397222           GLIWICE   
1          2016-01-01  00:05:00  19.308333  50.380833  DĄBROWA GÓRNICZA   
2          2016-01-01  18:30:00  18.976389  50.550833   TARNOWSKIE GÓRY   
3          2016-01-01  15:10:00  18.955833  50.546667   TARNOWSKIE GÓRY   
4          2016-01-01  17:45:00  19.175556  50.468056         OŻAROWICE   
...               ...       ...        ...        ...               ...   
229924     2019-12-10  10:30:00  19.228056  50.333056         SOSNOWIEC   
229925     2019-11-20  02:00:00  19.209722  50.008056         DANKOWICE   
229926     2019-11-24  14:30:00  19.311667  50.353056         SOSNOWIEC   
229927     2019-12-04  10:50:00  18.906389  49.927222          ZABŁOCIE   
229928     2019-04-07  19:25:00  19.123056  49.886667     BIELSKO-BIAŁA   

                       Ulica NrDomu    NrDrogi  KM HM (Pikietaż)  Obszar  ...  \
0                   TOSZECKA    NaN        901              61.6   False  ...   
1               SIENKIEWICZA    NaN  G200 030S               NaN    True  ...   
2          KS.IGNACEGO SIWCA     4A     GMINNA               NaN    True  ...   
3        JÓZEFA PIŁSUDSKIEGO      6     P3276S               NaN    True  ...   
4                   DWORCOWA     16     P3203S               NaN    True  ...   
...                      ...    ...        ...               ...     ...  ...   
229924  BRACI MIEROSZEWSKICH    NaN     P7605S               NaN    True  ...   
229925           OŚWIĘCIMSKA    NaN     P4444S               NaN    True  ...   
229926  BRACI MIEROSZEWSKICH    NaN     P7605S               NaN    True  ...   
229927               SKRAJNA      5   G611078S               NaN    True  ...   
229928           PARTYZANTÓW    156       W942               NaN    True  ...   

        Oznakowanie poziome Liczba pasów w zdarzeniu  \
0                      Jest                      2.0   
1                      Jest                      NaN   
2                    Nie ma                      2.0   
3                      Jest                      2.0   
4                    Nie ma                      2.0   
...                     ...                      ...   
229924                 Jest                      4.0   
229925                 Jest                      1.0   
229926                 Jest                      2.0   
229927               Nie ma                      2.0   
229928                 Jest                      2.0   

                             Zachowanie kierującego Zachowanie pieszego  \
0                                               NaN                 NaN   
1                            Nieprawidłowe: cofanie                 NaN   
2                           Nieprawidłowe: omijanie                 NaN   
3                           Nieprawidłowe: omijanie                 NaN   
4                                               NaN                 NaN   
...                                             ...                 ...   
229924          Nieprawidłowe zmienianie pasa ruchu                 NaN   
229925   Nieprzestrzeganie znaków i innych sygnałów                 NaN   
229926                   Nieprawidłowe wyprzedzanie                 NaN   
229927                                          NaN                 NaN   
229928  Niedostosowanie prędkości do warunków ruchu                 NaN   

                      Inne przyczyny Liczba rannych w zdarzeniu  \
0       Obiekty, zwierzęta na drodze                          0   
1                                NaN                          0   
2                                NaN                          0   
3                                NaN                          0   
4       Obiekty, zwierzęta na drodze                          0   
...                              ...                        ...   
229924                           NaN                          0   
229925                 

In [19]:
df['Skrzyżowanie']

0         NaN
1         2.0
2         NaN
3         NaN
4         NaN
         ... 
229924    NaN
229925    0.0
229926    2.0
229927    NaN
229928    NaN
Name: Skrzyżowanie, Length: 229929, dtype: float64